src/nvim/getchar.c

```diff
@@ -1577,729 +1577,868 @@ vungetc ( /* unget one character (can only be done once!) */
   old_mouse_col = mouse_col;
 }
 
-/// get a character:
-/// 1. from the stuffbuffer
-///    This is used for abbreviated commands like "D" -> "d$".
-///    Also used to redo a command for ".".
-/// 2. from the typeahead buffer
-///    Stores text obtained previously but not used yet.
-///    Also stores the result of mappings.
-///    Also used for the ":normal" command.
-/// 3. from the user
-///    This may do a blocking wait if "advance" is TRUE.
-///
-/// if "advance" is TRUE (vgetc()):
-///    really get the character.
-///    KeyTyped is set to TRUE in the case the user typed the key.
-///    KeyStuffed is TRUE if the character comes from the stuff buffer.
-/// if "advance" is FALSE (vpeekc()):
-///    just look whether there is a character available.
-///
-/// When "no_mapping" is zero, checks for mappings in the current mode.
-/// Only returns one byte (of a multi-byte character).
-/// K_SPECIAL and CSI may be escaped, need to get two more bytes then.
-static int vgetorpeek(int advance)
+// The "INSERT" message is taken care of here:
+//     If vgetorpeek() is about to return an ESC to exit insert mode, the
+//     message is deleted
+//     If it's not going to return an ESC but has already deleted the message,
+//     redisplay it. This can be the case if we waited on an insert-mode
+//     mapping that began with ESC, and cleared the mode to give the appearance
+//     of an instant reaction to the ESC.
+static void do_insert_message(const bool mode_deleted, const int c,
+                              const int advance)
 {
-  int c, c1;
-  int keylen;
-  char_u      *s;
-  mapblock_T  *mp;
-  mapblock_T  *mp2;
-  mapblock_T  *mp_match;
-  int mp_match_len = 0;
-  int timedout = FALSE;                     /* waited for more than 1 second
-                                                for mapping to complete */
-  int mapdepth = 0;                 /* check for recursive mapping */
-  int mode_deleted = FALSE;             /* set when mode has been deleted */
-  int local_State;
+  if (advance && p_smd && msg_silent == 0 && (State & INSERT)) {
+    if (c == ESC && !mode_deleted && !no_mapping && mode_displayed) {
+      if (typebuf.tb_len && !KeyTyped) {
+        redraw_cmdline = true;              // delete mode later
+      } else {
+        unshowmode(false);
+      }
+    } else if (c != ESC && mode_deleted) {
+      if (typebuf.tb_len && !KeyTyped) {
+        redraw_cmdline = true;              // show mode later
+      } else {
+        showmode();
+      }
+    }
+  }
+}
```

Comment:
This function touches many globals is deeply coupled with the current state of the world. It's called only once and there is no other scenario where this function would be called and do the right thing.